rskj-core/src/main/java/org/ethereum/db/IndexedBlockStore.java

```diff
@@ -135,6 +135,46 @@ public Block getBlockByHashAndDepth(byte[] hash, long depth) {
         return block;
     }
 
+    @Override
+    public Block getBlockAtDepthStartingAt(long depth, byte[] hash) {
+        Block start = this.getBlockByHash(hash);
+
+        if (start != null && depth == 0) {
+            return start;
+        }
+
+        if (start == null || start.getNumber() <= depth) {
+            return null;
+        }
+
+        Block block = start;
+
+        for (long i = 0; i < depth; i++) {
```

Comment:
I don't really understand this `for` loop. It doesn't even depend on the `i` value! Could you clarify please?

Reply:
It is an optimization to traverse a branch in search for a block at a given depth. Starting at a given block (by hash) it tries to find the first block that is part of the best chain, when it finds one we now that we can jump to the block that is at the remaining depth. If not block is found then it continues traversing the branch from parent to parent. The search is limited by the maximum depth received as parameter. The method either needs to traverse the parent chain or if a block in the parent chain is part of the best chain then it can skip the traversal by going directly to the block at the remaining depth. Could be rewritten as a while loop but it seems to me it will be less clear. Added this comment to the method as documentation.

Reply:
I don't understand why you have to traverse the branch. Is there anything wrong with just calling `getChainBlockByNumber`?

Reply:
The reason to do this is to verify that the block is part of the mainchain

Reply:
`getChainBlockByNumber` only returns blocks in the main chain

Reply:
Why not keeping a map from block hash to `Blockinfo` instead of a list?